### ISSUE TITLE
GridsearchCV score value

### ISSUE THREAD

**Issue by kmahjoory** [skorch-dev/skorch#862]
Hi, I am new to Braindecode and I am tuning hyperparameters using GridSearch. 
Below is the code I use:

`train_set = split['train']`
`train_X = SliceDataset(train_set, idx=0)`
`train_y = train_set.get_metadata().target`

`model = ShallowFBCSPNet(n_chans, n_classes, input_window_samples=input_window_samples, final_conv_length='auto',
)`

`clf = EEGClassifier(
module=model,
criterion=torch.nn.NLLLoss,
optimizer=torch.optim.AdamW,
batch_size=batch_size,
iterator_train__shuffle=True,
callbacks=["accuracy"])`

`param_grid = {'lr': [4*1e-3, 7*1e-3],
 'optimizer__weight_decay': [1e-4, 1e-3],
}`

`gs = GridSearchCV(clf, param_grid, refit=True, cv=5, verbose=3)`
`gs.fit(train_X, train_y)`

![image](https://user-images.githubusercontent.com/51201214/173554768-7d54b297-6e49-47bc-9e1f-ac1d25489d86.png)

shouldn't the score value obtained after testing the model on every fold be the same as the validation accuracy of the last epoch?
 

As far as I understood from codes, gs.fit calls the function _fit_and_score(), in which the train data is split into train and validation sets (say train1 and valid1) for cross validation. After that, NeuralNet.fit_loop() is called, where the train1 (but NOT train) is split once again into train and validation sets (say train2, and valid2). 
In the end, for every CV fold and every set of hyperparameters, the valid_accuracy (reported in the python console) is calculated from valid2 while the score value is calculated from valid1. Did I understand this correctly? If so, wondering why not the same validation data is used for calculating the valid_accuracy and score value?

Best,
Keyvan

**Comment by BenjaminBossan** [comment]
> As far as I understood from codes, gs.fit calls the function _fit_and_score(), in which the train data is split into train and validation sets (say train1 and valid1) for cross validation. After that, NeuralNet.fit_loop() is called, where the train1 (but NOT train) is split once again into train and validation sets (say train2, and valid2).
> In the end, for every CV fold and every set of hyperparameters, the valid_accuracy (reported in the python console) is calculated from valid2 while the score value is calculated from valid1. Did I understand this correctly?

Yes, you understand correctly, this is why the two scores are not exactly the same.

> If so, wondering why not the same validation data is used for calculating the valid_accuracy and score value?

Without going into too much detail, this is not really possible. The grid search is performed by sklearn and it purposefully does not pass the validation data in the fit step (to avoid any potential leaks). But skorch can only see the data made available for training, so it cannot evaluate the valid1 dataset.

There are possible workarounds to achieve the outcome you mentioned but I don't recommend them. At the end of the day, if the two exact same scores were calculated, they would be redundant. My proposal would be to completely turn off the skorch validation (train2) when you use grid search*, as the grid search already covers it. To do this, change the code to (I assume here that `EEGClassifier` works like a skorch `NeuralNet`):

```python
clf = EEGClassifier(..., train_split=False, verbose=0)
```

I also recommend to set `verbose=0` if you want to avoid too many print outputs.

*the main exception would be if you want to do early stopping on a held out dataset

**Comment by kmahjoory** [comment]
Thanks! 
Yes, exactly, the aim is to use early stopping on heald out data for gridsearchCV. In my current code, the accuracy is calculated from valid2 and is used  as a stopping criteria. The result of this, doesn't always match with the score value (obtained from valid1) . That's why I thought it would be nice to use valid1 for the early stopping as well.

**Comment by BenjaminBossan** [comment]
> the aim is to use early stopping

Personally, I consider data used for early stopping to be training data, even if it's not directly used to update the weights of the net. If you used the same data, you would actually peak (a little bit) at the validation data. Therefore, it's actually a good thing that it's not the same validation as used by the grid search.

**Comment by kmahjoory** [comment]
Yes, I do agree! thanks for sharing your points. That answers my question!
